crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -8,15 +8,68 @@
 //! reasonably ask us to type-check code assuming that the code runs
 //! on Linux.)
 
+use std::fmt;
 use std::io;
+use std::num::NonZeroUsize;
+use std::ops::Deref;
 
 use ruff_db::system::{System, SystemPath, SystemPathBuf};
 
-/// Attempt to retrieve the `site-packages` directory
-/// associated with a given Python installation.
+/// The value given by the `home` key in `pyvenv.cfg` files.
+///
+/// This is equivalent to `{sys_prefix_path}/bin`, and points
+/// to a directory in which a Python executable can be found.
+/// Confusingly, it is *not* the same as the [`PYTHONHOME`]
+/// environment variable that Python provides! However, it's
+/// consistent among all mainstream creators of Python virtual
+/// environments (the stdlib Python `venv` module, the third-party
+/// `virtualenv` library, and `uv`), was specified by
+/// [the original PEP adding the `venv` module],
+/// and it's one of the few fields that's read by the Python
+/// standard library's `site.py` module.
+///
+/// Other values, such as the path to the Python executable or the
+/// base-executable `sys.prefix` value, are either only provided in
+/// `pyvenv.cfg` files by some virtual-environment creators,
+/// or are included under different keys depending on which
+/// virtual-environment creation tool you've used.
+///
+/// [`PYTHONHOME`]: https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHOME
+/// [the original PEP adding the `venv` module]: https://peps.python.org/pep-0405/
+#[derive(Debug, PartialEq, Eq)]
+struct PythonHomePath(SystemPathBuf);
+
+impl PythonHomePath {
+    fn new(path: impl AsRef<SystemPath>, system: &dyn System) -> Option<Self> {
+        let path = path.as_ref();
+        // It's important to resolve symlinks here rather than simply making the path absolute,
+        // since system Python installations often only put symlinks in the "expected"
+        // locations for `home` and `site-packages`
+        let canonicalized = system.canonicalize_path(path).ok()?;
```

Comment:
can you explain why it is `ok` to return `None` if the canonicalize fails. Shouldn't we return an error in that case (or at least log a warning) saying that we skipped that path because we don't have access or the symlink is broken.

Reply:
I'm not sure I understand. If canonicalisation fails, we immediately return `None` here, indicating that we failed to construct a `PythonHomePath` or a `SysPrefixPath`. The `.ok()?` call ensures that. At both places where we call `PythonHomePath::new()` or `SysPrefixPath::new()`, `None` being returned is immediately converted into an error. So we're not silently ignoring any errors here if canonicalisation fails.

Would you prefer it if this method return a `Result` rather than an `Option`? Would that make it clearer, in your opinion?

Reply:
> Would you prefer it if this method return a Result rather than an Option? Would that make it clearer, in your opinion?

Returning a `Result` has the advantage that we can preserve the failure-cause which would be nice if this is an error that we show the user (or print to logs)